behaviour/graql/reasoner/recursion.feature

```diff
@@ -0,0 +1,2025 @@
+      (parent: $e, child: $d) isa Parent;
+
+      (parent: $f, child: $e) isa Parent;
+
+      #Extra data
+      (parent: $g, child: $f) isa Parent;
+
+      (parent: $h, child: $g) isa Parent;
+      """
+#    When materialised keyspace is completed
+    Then for graql query
+      """
+      match
+        ($x, $y) isa SameGen;
+        $x has name 'a';
+      get $y;
+      """
+#    Then all answers are correct in reasoned keyspace
+    Then answer size in reasoned keyspace is: 2
+    Then answer set is equivalent for graql query
+      """
+      match
+        $y has name $name;
+        {$name == 'f';} or {$name == 'a';};
+      get $y;
+      """
+#    Then materialised and reasoned keyspaces are the same size
+
+
+  # TODO: re-enable all steps when resolvable ('all answers correct' takes too long, 'same size' test fails) (#75)
+  Scenario: TC test
+
+    from Vieille - Recursive Query Processing: The power of logic p. 18
+
+    Given for each session, graql define
+      """
+      define
+
+      entity2 sub entity,
+          has index;
+      q sub entity2;
+
+      N-TC sub relation, relates N-TC-roleB, relates N-TC-roleA;
+      entity2 plays N-TC-roleB, plays N-TC-roleA;
+
+      TC sub relation, relates TC-roleA, relates TC-roleB;
+      entity2 plays TC-roleA, plays TC-roleB;
+
+      P sub relation, relates P-roleA, relates P-roleB;
+      entity2 plays P-roleA, plays P-roleB;
+
+      index sub attribute, value string;
+
+      rule-1 sub rule,
+      when {
+        $x isa q;
+        (TC-roleA: $x, TC-roleB: $y) isa TC;
+      }, then {
+        (N-TC-roleA: $x, N-TC-roleB: $y) isa N-TC;
+      };
+
+      rule-2 sub rule,
+      when {
+        (P-roleA: $x, P-roleB: $y) isa P;
+      }, then {
+        (TC-roleA: $x, TC-roleB: $y) isa TC;
+      };
+
+      rule-3 sub rule,
+      when {
+        (P-roleA: $x, P-roleB: $z) isa P;
+        (TC-roleA:$z, TC-roleB: $y) isa TC;
+      }, then {
+        (TC-roleA: $x, TC-roleB: $y) isa TC;
+      };
+      """
+    Given for each session, graql insert
+      """
+      insert
+
+      $a isa entity2, has index "a";
+      $a1 isa entity2, has index "a1";
+      $a2 isa q, has index "a2";
+
+      (P-roleA: $a1, P-roleB: $a) isa P;
+      (P-roleA: $a2, P-roleB: $a1) isa P;
+      """
+    When materialised keyspace is completed
+    Then for graql query
+      """
+      match
+        ($x, $y) isa N-TC;
+        $y has index 'a';
+      get $x;
+      """
+#    Then all answers are correct in reasoned keyspace
+    Then answer size in reasoned keyspace is: 1
+    Then answer set is equivalent for graql query
+      """
+      match $x has index 'a2'; get;
+      """
+#    Then materialised and reasoned keyspaces are the same size
+
+
+  # TODO: re-enable all steps when resolvable (currently takes too long) (#75)
+  Scenario: given a directed graph, all pairs of vertices (x,y) such that y is reachable from x can be found
+
+    test 5.2 from Green
```

Comment:
Green - Datalog and Recursive Query Processing

Reply:
Citation added